[css-masonry] Refactor Overriding of Logical Width during Masonry

stwrt opened the pull request:
#### 578b0cd616d27f530dd49e693bc7d6b4ac93fe4a
<pre>
[css-masonry] Refactor Overriding of Logical Width during Masonry
<a href="https://bugs.webkit.org/show_bug.cgi?id=282638">https://bugs.webkit.org/show_bug.cgi?id=282638</a>
<a href="https://rdar.apple.com/problem/139312026">rdar://problem/139312026</a>

Reviewed by Sammy Gill.

The overriding logical width logic should live in the GridMasonryLayout instead of RenderGrid.

Added an enum to handle the different min, max, and not overriding cases.

There was one special behavior with subgrid containers not being iterated over in the old implementation. I added extra code to handle this behavior.

 * Source/WebCore/rendering/GridMasonryLayout.cpp:
 (WebCore::GridMasonryLayout::performMasonryPlacement):
 (WebCore::GridMasonryLayout::placeItemsUsingOrderModifiedDocumentOrder):
 (WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::calculateMasonryIntrinsicLogicalWidth):
 (WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
 * Source/WebCore/rendering/GridMasonryLayout.h:
 * Source/WebCore/rendering/RenderGrid.cpp:
 (WebCore::RenderGrid::layoutMasonry):
 (WebCore::RenderGrid::computeIntrinsicLogicalWidths const):

Canonical link: <a href="https://commits.webkit.org/286276@main">https://commits.webkit.org/286276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14e92f49087841c1750ed0e94d0695fcd6fea6c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79837 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77480 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59146 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17374 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78431 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39508 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/10 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22242 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24955 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22580 "Found 1 new test failure: media/media-vp8-webm-with-poster.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2703 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1699 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67394 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64734 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66674 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16609 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10627 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8787 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2660 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5481 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2685 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3620 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2692 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -175,8 +201,9 @@ void GridMasonryLayout::setItemGridAxisContainingBlockToGridArea(const GridTrack
-void GridMasonryLayout::insertIntoGridAndLayoutItem(const GridTrackSizingAlgorithm& algorithm, RenderBox& gridItem, const GridArea& area)
+void GridMasonryLayout::insertIntoGridAndLayoutItem(const GridTrackSizingAlgorithm& algorithm, RenderBox& gridItem, const GridArea& area, GridMasonryLayout::OverrideLogicalWidth overrideLogicalWidthType)
 {
+    overrideLogicalWidth(gridItem, overrideLogicalWidthType);
```
imo it's not great to have a generic method like this that takes in a renderer and mutates it in such a way. I feel like it makes it easy to misuse. Why not just have this method compute a size and shift the responsibility to the caller to figure out what to do with it since they probably have more context as to what's going on.

stwrt replied:
That seems fine will update this to have the above function do the setting of logical widths.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,13 +154,39 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+        break;
+    case OverrideLogicalWidth::MaxContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent()) && !(gridItem.isOutOfFlowPositioned() || gridItem.isExcludedFromNormalLayout())) {
+            auto* renderGridMax = dynamicDowncast<RenderGrid>(gridItem);
```
What is a renderGridMax?

stwrt replied:
Update to just renderGrid

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,13 +154,39 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+        break;
+    case OverrideLogicalWidth::MinContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent()) && !(gridItem.isOutOfFlowPositioned() || gridItem.isExcludedFromNormalLayout())) {
+            auto* renderGridMin = dynamicDowncast<RenderGrid>(gridItem);
```
What is a renderGridMin?

stwrt replied:
Was running into compile errors when I had called it renderGrid in the past.

I tried it again and it compiled successfully.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.h:
```diff
@@ -41,8 +41,14 @@ class GridMasonryLayout {
     {
     }
 
+    enum class OverrideLogicalWidth {
```
Not 100% sure that `OverrideLogicalWidth` is a great name for this enum. Currently it implies that if you pass this into masonry layout some sort of logical width but not sure for what. Maybe `OVerrideMasonryLogicalWidth`?

stwrt replied:
Sounds good. Updated.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.h:
```diff
@@ -41,8 +41,14 @@ class GridMasonryLayout {
     }
 
+    enum class OverrideLogicalWidth {
+        DoNotOverride,
```
We could also just have the two `MinContent`/`MaxContent` values and pass in the enum as an optional but I'm not sure which makes more sense for this case or is better. Something to think about.

stwrt replied:
Sure, we could do something like that. Let me update this.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,13 +154,39 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+    case OverrideLogicalWidth::DoNotOverride:
+        break;
+    case OverrideLogicalWidth::MinContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent()) && !(gridItem.isOutOfFlowPositioned() || gridItem.isExcludedFromNormalLayout())) {
```
Can we not continue to reuse the order iterator logic for skipping grid items here? Seems like some of the other masonry code also uses it

stwrt replied:
These extra checks are not needed as we perform the checks earlier. Removed.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,13 +154,39 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+        }
+        break;
+    case OverrideLogicalWidth::MaxContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent()) && !(gridItem.isOutOfFlowPositioned() || gridItem.isExcludedFromNormalLayout())) {
```
Same question about the order iterator here

stwrt replied:
Removed the extra check here; see comment above.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,16 +154,40 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+    case OverrideMasonryLogicalWidth::MinContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent())) {
+            if (auto* renderGrid = dynamicDowncast<RenderGrid>(gridItem); renderGrid && renderGrid->isSubgridRows())
+                break;
```
Since this logic is repeated in both cases can we just pull it out of the switch statement and put it at the beginning?

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,16 +154,40 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+{
+    switch (overrideLogicalWidthType) {
+    case OverrideMasonryLogicalWidth::MinContent:
+        if ((gridItem.style().logicalWidth().isAuto() || gridItem.style().logicalWidth().isPercent())) {
```
Maybe also consider pulling this to the top as a way to return as early as possible if this condition is not true?

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -175,8 +199,11 @@ void GridMasonryLayout::setItemGridAxisContainingBlockToGridArea(const GridTrack
-void GridMasonryLayout::insertIntoGridAndLayoutItem(const GridTrackSizingAlgorithm& algorithm, RenderBox& gridItem, const GridArea& area)
+void GridMasonryLayout::insertIntoGridAndLayoutItem(const GridTrackSizingAlgorithm& algorithm, RenderBox& gridItem, const GridArea& area, std::optional<GridMasonryLayout::OverrideMasonryLogicalWidth> overrideLogicalWidthType)
 {
+    if (overrideLogicalWidthType.has_value())
```
Nit:
```suggestion
    if (overrideLogicalWidthType)
```

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,16 +154,40 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+            if (auto* renderGrid = dynamicDowncast<RenderGrid>(gridItem); renderGrid && renderGrid->isSubgridRows())
+                break;
+
+            return gridItem.computeIntrinsicLogicalWidthUsing(Length(LengthType::MinContent), LayoutUnit(), gridItem.borderAndPaddingLogicalWidth());
```
Nit: can replace `LayoutUnit()` with { }

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -154,16 +154,40 @@ GridArea GridMasonryLayout::gridAreaForDefiniteGridAxisItem(const RenderBox& gri
+            if (auto* renderGrid = dynamicDowncast<RenderGrid>(gridItem); renderGrid && renderGrid->isSubgridRows())
+                break;
+
+            return gridItem.computeIntrinsicLogicalWidthUsing(Length(LengthType::MaxContent), LayoutUnit(), gridItem.borderAndPaddingLogicalWidth());
```
Nit: can replace `LayoutUnit()` with { }

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.h:
```diff
@@ -41,8 +41,13 @@ class GridMasonryLayout {
     {
     }
 
+    enum class OverrideMasonryLogicalWidth {
```
```suggestion
    enum class OverrideMasonryLogicalWidth : bool {
```

stwrt committed:
[css-masonry] Refactor Overriding of Logical Width during Masonry
https://bugs.webkit.org/show_bug.cgi?id=282638
rdar://problem/139312026

Reviewed by Sammy Gill.

The overriding logical width logic should live in the GridMasonryLayout instead of RenderGrid.

Added an enum to handle the different min, max, and not overriding cases.

There was one special behavior with subgrid containers not being iterated over in the old implementation. I added extra code to handle this behavior.

 * Source/WebCore/rendering/GridMasonryLayout.cpp:
 (WebCore::GridMasonryLayout::performMasonryPlacement):
 (WebCore::GridMasonryLayout::placeItemsUsingOrderModifiedDocumentOrder):
 (WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::calculateMasonryIntrinsicLogicalWidth):
 (WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
 * Source/WebCore/rendering/GridMasonryLayout.h:
 * Source/WebCore/rendering/RenderGrid.cpp:
 (WebCore::RenderGrid::layoutMasonry):
 (WebCore::RenderGrid::computeIntrinsicLogicalWidths const):

Canonical link: https://commits.webkit.org/286276@main
